micro-rdk/src/esp32/nvs_storage.rs

```diff
@@ -126,6 +126,40 @@ impl NVSStorage {
     }
 }
 
+impl StorageDiagnostic for NVSStorage {
+    type Error = NVSStorageError;
+    fn log_space_diagnostic(&self) -> Result<(), Self::Error> {
+        let part_name = CString::new("nvs")
+            .map_err(|_| EspError::from_non_zero(NonZeroI32::new(ESP_ERR_INVALID_ARG).unwrap()))?;
+        let mut stats: nvs_stats_t = Default::default();
+        esp!(unsafe { nvs_get_stats(part_name.as_ptr(), &mut stats as *mut _) })?;
+
+        // 32 bytes per entry
+        let used_entries = stats.used_entries;
+        let used_space = used_entries * 32;
+        let used_page_header_space = (used_entries % 125) * 32;
+
+        let used_space = used_space + used_page_header_space;
+        let total_space = stats.total_entries * 32;
+
+        log::info!("CREDENTIAL STORAGE STATS: {:?} bytes used", used_space);
+        // From experimentation we have found that NVS requires 4000 bytes of
+        // unused space for reasons unknown. The percentage portion of the calculation
+        // comes from the maximum allowed blob size stated in the ESP32 documentation
+        // on NVS
+        let total_usable_space = (0.976 * (total_space as f64)) - 4000.0;
```

Comment:
why `0.976`? just to have room before panicking or something?
also, is this correct that we want to subtract 4000 from this percentage of total space? 
it's not `total_usable_space = ((total_space as f64) - 4000.0)`

Reply:
It was taken from the length restrictions here: https://docs.espressif.com/projects/esp-idf/en/v4.4.8/esp32/api-reference/storage/nvs_flash.html#keys-and-values